CORDA-1349: Docs: improve docs on permissioning, doorman and network map.

mikehearn opened the pull request.

fenryka commented on docs/source/permissioning.rst:
```diff
@@ -61,31 +53,26 @@ We can visualise the permissioning structure as follows:
-* The root network CA, doorman CA and node CA keys, as well as the node TLS
-  keys, must follow one of the following schemes:
-
+1. The certificates must follow the `X.509v3 standard <https://tools.ietf.org/html/rfc5280>`_
```
I have memories of it being mentioned a double trailing underscore was better... but I could be imagining that

fenryka commented on docs/source/permissioning.rst:
```diff
@@ -61,31 +53,26 @@ We can visualise the permissioning structure as follows:
-
+1. The certificates must follow the `X.509v3 standard <https://tools.ietf.org/html/rfc5280>`_
+2. The TLS certificates must follow the `TLS v1.2 standard <https://tools.ietf.org/html/rfc5246>`_
+3. The root network CA, doorman CA and node CA keys, as well as the node TLS keys, must follow one of the following schemes:
```
will always express desire for an Oxford comma

mikehearn replied:
Hmm I was always taught in school *not* to put a comma at the end of a list like that. At any rate either is correct, so I'll leave it.

fenryka commented on docs/source/permissioning.rst:
```diff
@@ -61,31 +53,26 @@ We can visualise the permissioning structure as follows:
-.. note:: Corda's ``X509Utilities`` show how to generate the required public/private keypairs and certificates using
-   Bouncy Castle. You can find the ``X509Utilities`` in the `Corda repository <https://github.com/corda/corda>`_, under
-   ``/node-api/src/main/kotlin/net/corda/nodeapi/internal/crypto/X509Utilities.kt``.
+The certificates and keys will be automatically generated for you by the node on first run. But you can also generate
```
However, you can also...

gendal replied:
Which certificates and keys will be auto generated?

fenryka commented on docs/source/permissioning.rst:
```diff
@@ -97,98 +84,36 @@ The extension contains a single ASN.1 integer identifying the identity type the
 To connect to a compatibility zone you need to register with their certificate signing authority (doorman) by submitting
-a certificate signing request (CSR) to obtain a valid identity for the zone.
+a certificate signing request (CSR) to obtain a valid identity for the zone. You could do this out of band, for instance
+via email or a web form, but there's also a simple request/response protocol built in to Corda.
```
into not in to (I think)

fenryka commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+
+This step is optional because your users can obtain a signed certificate in many different ways. The doorman protocol
+is again a very simple HTTP based approach in which a node creates keys and requests a certificate, polling until it
+gets back what it expects. But you could also integrate this process with the rest of your signup process, for example,
```
However, not but

fenryka replied:
<full stop>For example,

fenryka commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+Signing and saving the resulting file is just a few lines of code. A full example can be found in ``NetworkParametersCopier.kt`` in the
+source tree, but a flavour of it looks like this:
+
+.. sourcecode:: kotlin
```
Should we provide a Java one as well. Know Richard is keen we lean more on that to avoid the impression you have to use kotling for Corda

gendal replied:
good point...!!! Java first!!

fenryka commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+
+Once saved, the new parameters can be served by the network map server.
+
+How to choose the parameters? This is the most complex question facing you as a new zone operator. Some settings may seem
```
Feels like this might be a sub heading

gendal commented on docs/source/permissioning.rst:
```diff
@@ -3,55 +3,47 @@ Network permissioning
-
-.. note:: If you are looking for information on how to connect to the existing compatibility zone go to the section: `Connecting to a compatibility zone`_
+Production deployments require a secure certificate authority. Most users will join an existing network such as the
+main Corda network or the Corda TestNet. You can also build your own networks. During development, no particular
```
i think we decided not to capitalise the N in Testnet.

gendal commented on docs/source/permissioning.rst:
```diff
@@ -3,55 +3,47 @@ Network permissioning
 
-   * Node's well-known legal identity, apart from signing transactions, can also issue certificates for confidential legal identities
+* Doorman certificates are issued by a network root. Network root certs do not contain a role extension.
+* Node certificates are issued by an entity with a doorman certificate.
```
what does it mean to be an entity "with" a doorman certificate? The certificate is marked as having a doorman role? How is this manifested?

gendal commented on docs/source/permissioning.rst:
```diff
@@ -3,55 +3,47 @@ Network permissioning
+the identity service and therefore there is no advantage in enforcing a specific structure on those certificates. The
+certificate hierarchy consistency checks are required because nodes can issue their own certificates and can set
+their own role flags on certificates, and it's important to verify that these are set consistently with the
+certificate hierarchy design. As as side-effect this also acts as a secondary depth restriction on issued
```
As a

gendal commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
 .. note:: You can exit the utility at any time if the approval process is taking longer than expected. The request process will resume on restart.
```
provided you restart with the same command-line flag?  Or does a node started without the --initial-registration flag know that it was in the middle of a request and resume its wait?

gendal commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+2. It isn't necessary to add another layer of permissioning or 'know your customer' requirements onto your app.
+
+**Testing.** Creating a production-ready zone isn't necessary for testing as you can use the *network bootstrapper*
+tool to create all the certificates, keys and distribute map files needed to run many nodes. The bootstrapper can
```
Oxford comma would definitely help here....  certificates, keys, and distribute...

gendal commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+create a network locally on your desktop/laptop but it also knows how to automate cloud providers via their APIs and
+using Docker. In this way you can bring up a simulation of a real Corda network with different nodes on different
+machines in the cloud for your own testing. Testing this way has several advantages, most obviously that you avoid
+flakyness caused by synchronisation issues during network bringup. You can read more about the reasons for the
```
can we reword this? inattentive readers will interpret this as our saying corda is flaky.  How about: "... most obviously that you avoid the need to wait for information about all nodes to propagate to all other nodes as they start up before being able to begin testing; instead, with this approach, you know for sure that when a node starts it has all the information about the network it needs in order to support testing immediately.

gendal commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+   members and at the start of each flow, throw a FlowException if the counterparty isn't in the membership list.
+
+In this way you can impose a centrally controlled ACL that all members will collectively enforce.
+
```
reiterate at this point that the production Corda network and Corda Testnet will shortly be available?

gendal commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+Writing a doorman server
+^^^^^^^^^^^^^^^^^^^^^^^^
+
+This step is optional because your users can obtain a signed certificate in many different ways. The doorman protocol
```
And one such way is from the Corda Network or Corda Testnet doorman, right?  You've argued at this point that the only reason to go down this path is because people want different network params but I think there will also be people who are quite happy to adopt OUR params but who aren't _quite_ ready to join CN/CT yet.  In such cases, do we want to recommend a path that says: "If you really, REALLY, REALLY must run your own private network, be aware that you are, by definition, isolated from the main network and will soon wish you'd chosen to join the big party.  To reduce the risk of encountering this remorse, we suggest adopting the CN/CT netparams and using certificates issued by one of those networks.  This doesn't guarantee you'll subsequently be able to join the real networks (not least because they need to have run their notaries to the right standards) but there's at least hope...  whereas there is no hope if they use different params or issue their own certs

gendal replied:
(but it does imply a small product change, I think?  since the private netmap server wouldn't be able to sign the network map it serves to the private network with the key that is associated with the CN/CT-issued certs)

gendal commented on docs/source/permissioning.rst:
```diff
@@ -221,3 +147,157 @@ Once the request has been approved and the certificates downloaded from the serv
+
+How to choose the parameters? This is the most complex question facing you as a new zone operator. Some settings may seem
+straightforward and others may involve cost/benefit tradeoffs specific to your business. For example, you could choose
+to run a validating notary yourself, in which case you would (in the absence of SGX) see all the users data. Or you could
```
users'

mikehearn commented:
Thanks both. I've addressed the review comments, with the exception of Richard's greater point.

I'm asking about the extent to which we can mix and match certs before things break. We didn't design for that and it'd probably complicate at least the security analysis.

shamsasari commented on docs/source/permissioning.rst:
```diff
@@ -210,14 +143,188 @@ The following information from the node configuration file is needed to generate
+Do you need a zone?
+^^^^^^^^^^^^^^^^^^^
+
+Think twice before going this route:
```
going down

shamsasari commented on node-api/src/main/kotlin/net/corda/nodeapi/internal/network/NetworkParametersCopier.kt:
```diff
@@ -19,7 +21,7 @@ class NetworkParametersCopier(
 ) {
     private val copyOptions = if (overwriteFile) arrayOf(StandardCopyOption.REPLACE_EXISTING) else emptyArray()
-    private val serialisedSignedNetParams = signingCertAndKeyPair.sign(networkParameters).serialize()
+    private val serialisedSignedNetParams: SerializedBytes<SignedDataWithCert<NetworkParameters>> = signingCertAndKeyPair.sign(networkParameters).serialize()
```
Use the `SignedNetworkParameters` typealias

davejh69 commented on docs/source/permissioning.rst:
```diff
@@ -210,14 +143,188 @@ The following information from the node configuration file is needed to generate
 
 This process only is needed when the node connects to the network for the first time, or when the certificate expires.
+
+Creating your own compatibility zone
```
This section seems a little awkward. We've started by describing the technical implementation of our certificate hierarchy, and how it fits into a CZ. This feels very much orthogonal to that and it veers towards being opinionated in an otherwise non-opinionated page. Feels like this should be a separate discussion. There are clear examples of places where a new CZ is required, even if they're not required for many of the reasons people might have expected.

mikehearn replied:
Yeah, I agree it's a bit incongruous, but we *are* opinionated about this. We really want everyone to be able to connect to each other!

davejh69 replied:
agreed that this is actually an opinion we want to express, but separating the technical content from the discursive would seem like a good idea. If anything it feels like we probably want to deal with the advocacy aspect up front. We describe the relationship of a node to its CZ, describe that there's a top-of-stack solution to support multiple BNs in a CZ, and then strongly suggest the use of tCN as a default CZ. If we then go into the identity hierarchy and what's required to manage that, we've already set an expectation that "default is tCN" and everything else is as much about understanding how that works as anything else (oh and, by the way, if you need the non-default then this is what you'll need to do)

mikehearn replied:
I took another look at this, but, I can't quite convince myself that it's right to move this discussion to the top of the page. It's perfectly reasonable to want to understand permissioning *without* wanting to read about creating or not creating new zones. And I don't want to put the discussion of whether to do it in a separate page that risks people not reading it.

In conclusion I'd rather leave it as it is, though, I'd be happy to review PRs that have a concrete proposal for changing the layout or language.

davejh69 commented on docs/source/permissioning.rst:
```diff
@@ -97,98 +91,36 @@ The extension contains a single ASN.1 integer identifying the identity type the
-In a typical installation, node administrators needn't be aware of these. However, when node certificates are managed
+In a typical installation, node administrators need not be aware of these. However, when node certificates are managed
 by external tools (such as an existing PKI solution deployed within an organisation), it is important to understand
 these constraints.
```
what constraints? We've not describe that anything is being constrained? If the certs are being constrained by the extension then what must an otherwise-valid cert not contain?

mikehearn replied:
We've described the name constraints and the extension that defines the roles; that's what this is referring to. Can you see a way to make that sentence clearer?

davejh69 replied:
How about: "In a typical installation, node administrators need not be aware of these. However, if node certificates are to be managed by external tools, such as those provided as part of an existing PKI solution deployed within an organisation, it is important to recognise these extensions and the constraints noted above."

mikehearn commented:
Can I get an approval from someone?

gendal commented on docs/source/permissioning.rst:
```diff
@@ -268,14 +274,14 @@ The protocol is:
+object. It is easy to create such a file with a small Java or Kotlin program. The ``NetworkParameters`` object is a
+simple data holder that could be read from e.g. a config file, or settings from a database. Signing and saving the
+resulting file is just a few lines of code. A full example can be found in ``NetworkParametersCopier.kt`` in the source
+tree, but a flavour of it looks like this:
```
whenever I see something like this I think... "a vendor of software is telling me that doing thing X is a simple matter of writing some code... but isn't that the vendor's job?!!  Can we perhaps file a Jira (it's not a blocker to this PR) to create a trivial tool that does just this?

mikehearn replied:
https://r3-cev.atlassian.net/browse/CORDA-1592

gendal replied:
👍
